File: frontend/src/types/coupon.ts
export type DiscountType = "percentage" | "fixed"

export interface CouponBase {
  code: string
  discount_type: DiscountType
  discount_value: number
  min_order_value?: number | null
  max_discount?: number | null
  is_active: boolean
  usage_limit?: number | null
  valid_from?: string | null
  valid_until?: string | null
}

export interface CouponOut extends CouponBase {
  id: string
  usage_count: number
  created_at: string
  updated_at: string
}

export type CouponCreate = CouponBase

export interface CouponUpdate {
  code?: string
  discount_type?: DiscountType
  discount_value?: number
  min_order_value?: number | null
  max_discount?: number | null
  is_active?: boolean
  usage_limit?: number | null
  valid_from?: string | null
  valid_until?: string | null
}

export interface CouponValidateRequest {
  code: string
  cart_total: number
}

export interface CouponValidateResponse {
  is_valid: boolean
  discount_amount: number
  message?: string
  coupon?: CouponOut
}

File: backend/app/schemas/coupon.py
from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, Field, constr

from app.models.coupon import DiscountType


class CouponBase(BaseModel):
    code: str = Field(..., max_length=50)
    discount_type: DiscountType = Field(default=DiscountType.percentage)
    discount_value: float = Field(..., gt=0)
    
    min_order_value: float | None = Field(default=None, ge=0)
    max_discount: float | None = Field(default=None, ge=0)
    
    is_active: bool = True
    usage_limit: int | None = Field(default=None, ge=1)
    
    valid_from: datetime | None = None
    valid_until: datetime | None = None


class CouponCreate(CouponBase):
    pass


class CouponUpdate(BaseModel):
    code: str | None = Field(default=None, max_length=50)
    discount_type: DiscountType | None = None
    discount_value: float | None = Field(default=None, gt=0)
    min_order_value: float | None = Field(default=None, ge=0)
    max_discount: float | None = Field(default=None, ge=0)
    is_active: bool | None = None
    usage_limit: int | None = Field(default=None, ge=1)
    valid_from: datetime | None = None
    valid_until: datetime | None = None


class CouponOut(CouponBase):
    id: UUID
    usage_count: int
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class CouponValidateRequest(BaseModel):
    code: str
    cart_total: float = Field(..., ge=0)


class CouponValidateResponse(BaseModel):
    is_valid: bool
    discount_amount: float
    message: str | None = None
    coupon: CouponOut | None = None
